strawberryfields/engine.py

```diff
@@ -531,161 +544,541 @@ def run(self, program, *, args=None, compile_options=None, run_options=None):
     """
 
-    # This engine will execute jobs remotely.
-    REMOTE = True
+    def __init__(self, id_: str, status: JobStatus, connection: "Connection"):
+        self._id = id_
+        self._status = status
+        self._connection = connection
+        self._result = None
+
+    @property
+    def id(self) -> str:
+        """The job ID.
+
+        Returns:
+            str: the job ID
+        """
+        return self._id
+
+    @property
+    def status(self) -> JobStatus:
+        """The job status.
+
+        Returns:
+            strawberryfields.engine.JobStatus: the job status
+        """
+        return self._status
+
+    @property
+    def result(self) -> Result:
+        """The job result.
 
-    def __init__(self, backend, polling_delay_seconds=1, **kwargs):
-        super().__init__(backend)
+        This is only defined for complete jobs, and raises an exception for any other
+        status.
 
-        api_client_params = {k: v for k, v in kwargs.items() if k in DEFAULT_CONFIG["api"].keys()}
-        self.client = APIClient(**api_client_params)
-        self.polling_delay_seconds = polling_delay_seconds
-        self.jobs = []
+        Returns:
+            strawberryfields.engine.Result: the job result
+        """
+        if self.status != JobStatus.COMPLETE:
+            raise AttributeError(
+                "The result is undefined for jobs that are not complete "
+                "(current status: {})".format(self.status.value)
+            )
+        return self._result
+
+    def refresh(self):
+        """Refreshes the status of the job, along with the job result if the job is
+        newly completed.
+
+        Only an open or queued job can be refreshed; an exception is raised otherwise.
+        """
+        if self.status.is_terminal:
+            raise InvalidJobOperationError(
+                "A {} job cannot be refreshed".format(self.status.value)
+            )
+        self._status = self._connection.get_job_status(self.id)
+        if self._status == JobStatus.COMPLETE:
+            self._result = self._connection.get_job_result(self.id)
+
+    def cancel(self):
+        """Cancels the job.
+
+        Only an open or queued job can be cancelled; an exception is raised otherwise.
+        """
+        if self.status.is_terminal:
+            raise InvalidJobOperationError(
+                "A {} job cannot be cancelled".format(self.status.value)
+            )
+        self._connection.cancel_job(self.id)
+
+    def __repr__(self):
+        return "<{}: id={}, status={}>".format(
+            self.__class__.__name__, self.id, self.status.value
+        )
 
     def __str__(self):
-        return self.__class__.__name__ + "({})".format(self.backend_name)
+        return self.__repr__()
+
+
+class RequestMethod(enum.Enum):
+    """Defines the valid request methods for messages sent to the remote platform."""
+
+    GET = "get"
+    POST = "post"
+    PATCH = "patch"
+
+
+class Connection:
```

Comment:
Hmm, this file seems to be getting quite big, maybe it would be worth having the `Connection` class, `Job` and related `Error` classes in a separate one called `connection.py`? :thinking: Or to split them is some other meaningful way :blush:

Reply:
Great idea! Especially as the `Connection`/`Job` interface is intended to be usable independent of the `StarshipEngine`, and the module boundary would help to enforce and make explicit that separation.

One problem I see though is that creates a circular dependency due to the `engine.Result` class, which would to be imported in `connection.py`---and `engine.py` would need to import `Connection` from `connection.py`, etc.

A solution to that might be to move `Result` into a separate module. What do you think?

Reply:
Sounds good to me :)

Reply:
:+1: I commented the exact same thing last night, but couldn't submit my review because github buttons stopped working 😆 But yes, I agree. Options include:

* Package: `api/connection.py`, `api/job,py`, etc.

* A module `connection.py` that contains all classes

I tend to prefer option (1), an API package, as it collects all objects as the same level of abstraction.

Reply:
I've moved the relevant code into the following modules: `api/connection.py`, `api/job.py`, and `api/result.py`.